File: goldeneye/src/main/java/co/infinum/goldeneye/models/Size.java
package co.infinum.goldeneye.models;

import android.hardware.Camera;
import android.support.annotation.NonNull;

public class Size implements Comparable<Size> {

    public static final Size UNKNOWN = new Size(0, 0);
    private static final float EPSILON = 0.0001f;
    private final int height;
    private final int width;

    public Size(int width, int height) {
        this.width = width;
        this.height = height;
    }

    public Size(Camera.Size size) {
        this(size.width, size.height);
    }

    @Override
    public int compareTo(@NonNull Size o) {
        return o.height * o.width - width * height;
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof Size
            && ((Size) obj).height == height
            && ((Size) obj).width == width;
    }

    public float getAspectRatio() {
        return ((float) width) / height;
    }

    public int getHeight() {
        return height;
    }

    public int getWidth() {
        return width;
    }

    public boolean isSameAspectRatio(Size other) {
        return Float.compare(getAspectRatio(), other.getAspectRatio()) == 0;
    }
}

File: goldeneye/src/main/java/co/infinum/goldeneye/models/Size.kt
@file:Suppress("DEPRECATION")

package co.infinum.goldeneye.models

import android.hardware.Camera
import android.os.Build
import android.support.annotation.RequiresApi

data class Size internal constructor(
    val width: Int,
    val height: Int
) : Comparable<Size> {
    companion object {
        val UNKNOWN = Size(0, 0)
    }

    override fun compareTo(other: Size): Int {
        return other.height * other.width - width * height
    }

    val aspectRatio = if (this.height != 0 && this.width != 0) this.width.toFloat() / this.height.toFloat() else -1f
    fun isOver1080p() = this.width > 1080 && this.height > 1080
}

internal fun Camera.Size.toInternalSize() = Size(width, height)

@RequiresApi(Build.VERSION_CODES.LOLLIPOP)
internal fun android.util.Size.toInternalSize() = Size(width, height)